### PR TITLE
[move-prover] Fixing issues with reference parameters for function values

### LINKED ISSUE #17420: [Bug][Prover] reference type is not supported for function value types during generation of boogie code
# 🐛 Bug

When generating data type for function types, parameter types are used as suffix in the type name. However, if the function type contains reference as parameters such as:

```
module 0x42::test {


    fun exec(f: |&u64|u64, x: &u64): u64 {
        f(x)
    }

    fun call_exec1_ok(x: u64, z: &u64): u64 {
        exec(|y| x + *y, z)
    }
}

```

boogie will fail due to generated string "<<unsupported Reference(Immutable, Primitive(U64))>>".

### PR DESCRIPTION
**Pull request by wrwg**
## Description

This fixes #17420 and additional problems with function values which take references as parameters. See the test cases for what is working now.

There is one open problem which has been documented in #17904. This case passes compilation now, just cannot be verified. (It's hopefully a corner case for which we don't need a solution right away.)

This also improves error message on function value use in spec expressions, which is still not supported.

## How Has This Been Tested?

New tests

## Type of Change
- [ ] New feature
- [x] Bug fix
- [ ] Breaking change
- [ ] Performance improvement
- [ ] Refactoring
- [ ] Dependency update
- [ ] Documentation update
- [ ] Tests

## Which Components or Systems Does This Change Impact?
- [ ] Validator Node
- [ ] Full Node (API, Indexer, etc.)
- [ ] Move/Aptos Virtual Machine
- [ ] Aptos Framework
- [ ] Aptos CLI/SDK
- [ ] Developer Infrastructure
- [x] Move Compiler
- [ ] Other (specify)

### CODE REVIEW

**Comment by wrwg** [comment]
* **#17907** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/17907?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/17907?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: third_party/move/move-model/bytecode/src/borrow_analysis.rs [hunk 1]
```diff
@@ -240,10 +240,12 @@ impl BorrowInfo {
             "borrowed_by",
             self.borrowed_by.iter().map(borrows_str).join(", "),
         );
+        /*
```
Comment (rahxephon89) [inline third_party/move/move-model/bytecode/src/borrow_analysis.rs]: Is this intended?

Comment (wrwg) [reply third_party/move/move-model/bytecode/src/borrow_analysis.rs]: Ooops, restored.
